### DISCUSSION TITLE
Return an error message to client in the same page

### DISCUSSION THREAD

**Discussion by natsathorn** [nextauthjs/next-auth#754] [Help] [answered]
**Your question**
How to return the message to the login form?

**What are you trying to do**
I'm using Credentials provider which will just fire a request to another server for login. It's work fine when everything go as plan. In the case that login is not success, I can only send the error message via querystring which make the page reloaded.
Is there any other way to send the errorMessage back to the form without reload it.

FYI. I used `signIn('myAuth', { ...data })` on my login form.

**Feedback**
In this [section](https://next-auth.js.org/providers/credentials#example), it only shows these two options.
```
// You can also Reject this callback with an Error or with a URL:
return Promise.reject(new Error('error message')) // Redirect to error page
return Promise.reject('/path/to/redirect')        // Redirect to a URL
```

* [ ] Found the documentation helpful
* [ ] Found documentation but was incomplete
* [x] Could not find relevant documentation
* [ ] Found the example project helpful
* [x] Did not find the example project helpful

**Comment by bdub9** [comment]
I'm having the same issue.  Desired use case to skip a redirect:

1. Check if a user/password combo exists or password match.
2. They don't. Would like to Promise.reject with an error code or error message WITHOUT redirecting. Doesn't seem like there's a way to skip the redirect, even Promise.reject('/login?errorMessage=tryAgain') here (my form is on /login) reloads the page. How can we just send down a simple error message in a response so our client can handle it?

**Comment by mdesilva** [comment]
I'm having the same issue as well. Would be nice if we can just do the login RESTfully and do the redirection ourselves.

**Comment by MateuszWawrzynski** [comment]
*bump*
@iaincollins 

I'm looking for the answer for exactly the same problem.
Any ideas?

Is there any way to handle error code on the signin page without redirecting to "error" page and just show the error under the login form?

**Comment by mdesilva** [comment]
You can return the error as a query parameter to the login page

**Comment by flosrn** [comment]
I would like to know if is there an other way to send error message to client besides with query params ? 
Because query params cause the refresh of entire page and cause flashing who is little bit weird for the client

**Comment by balazsorban44** [comment, marked as answer]
Hi everyone, I am working on this right now! 
Would this PR solve your issue?: #1219

**Reply by dovalid** [reply]
Yep, looks perfect! Thanks for your work.

**Comment by yannicktian** [comment]
Thanks @balazsorban44
As @StefanSelfTaught mentioned here, it would be super to have the option to disable redirect for Email provider as well
https://github.com/nextauthjs/next-auth/pull/1219#issuecomment-774483051

**Reply by yannicktian** [reply]
@balazsorban44 I have a small PR ready but cannot push
```
ERROR: Permission to nextauthjs/next-auth.git denied to $ME.
fatal: Could not read from remote repository.
```

Do I have to be added to a specific team or am I missing something 🤔 ?

**Reply by StefanSelfTaught** [reply]
@yannicktian  Did you fork the repo?

**Reply by yannicktian** [reply]
@StefanSelfTaught I've cloned it
`git clone git@github.com:nextauthjs/next-auth.git && cd next-auth`
- Pulling is ok, but I can't push

**Reply by balazsorban44** [reply]
You don't have push access to the repo, only maintainers do.

Check out this tutorial: https://jarv.is/notes/how-to-pull-request-fork-github/

**Reply by yannicktian** [reply]
@balazsorban44 @StefanSelfTaught Cheers sorry for this extremely noob discussion 😅 I should contribute more often to open source project, here is the PR https://github.com/nextauthjs/next-auth/pull/1379
